Proposing identifiers.

svenboulanger opened the pull request:
Proposal for allowing more freedom involving identifiers.

marcin-golebiowski commented on SpiceSharp/Simulations/Variables/VariableSet.cs:
```diff
@@ -56,8 +56,29 @@ public VariableSet()
+        /// Initializes a new instance of the <see cref="VariableSet"/> class.
+        /// </summary>
+        /// <param name="comparer">The comparer for identifiers.</param>
+        public VariableSet(IEqualityComparer<Identifier> comparer)
```
@svenboulanger  And I think it should have ...

svenboulanger commented:
I am probably going to avoid specifying these things in the constructor of `Simulation`. Instead, I'll add an *optional* configuration that allows you to specify the equality comparer.

We can do that because the objects in question are only built in the `Setup()` method, at that point we can read the configuration and build the lists.

svenboulanger commented on SpiceSharp/Circuits/Circuit.cs:
```diff
@@ -41,6 +41,16 @@ public class Circuit
+        /// Initializes a new instance of the <see cref="Circuit"/> class.
+        /// </summary>
+        /// <param name="comparer">The comparer for identifiers.</param>
+        public Circuit(IEqualityComparer<Identifier> comparer)
```
I have done this by adding the `CollectionsConfiguration` class which can be added in the `Configuration` of the simulation.

svenboulanger commented:
There are also other issues. Some behaviors create a new variable/node where they use `new SubIdentifier(..., "branch")`. In order for the second parameter to use a custom identifier, we need some kind of factory.

svenboulanger commented:
We can take this idea further and also add a similar construction for other implicit conversion types ( eg. `int` and/or `long`). What do you think? @marcin-golebiowski

marcin-golebiowski commented:
@svenboulanger
I think factory idea with static method will not allow me to introduce different case-sensitivity settings for nodes, entities, parameters, function names. Maybe that feature is not really needed but will make the parser more powerful.

marcin-golebiowski commented:
@svenboulanger  Maybe we could combine two ideas.
I could introduce "Default setting" for everything else that will be controlled with that static factory and everything else with proper comparers. I'm not sure.

svenboulanger commented:
@marcin-golebiowski Maybe we are over-designing identifiers? We can also just move to plain strings. Simple and easy.

marcin-golebiowski commented:
@svenboulanger  I like the idea 👍 I vote for no implicit conversions. However the comparers should stay.

svenboulanger commented:
@marcin-golebiowski I removed `Identifier` and replaced everything with normal strings.
One note though:
Some behaviors create an extra variable that has to be named. I made this somewhat configurable by adding the static `Combine` method, which just concatenates strings with an optional separator.

marcin-golebiowski commented:
Works like charm :) 
However some other changes are required (see pull request #43)

svenboulanger commented:
I'll go ahead and merge this into the development branch.

marcin-golebiowski commented:
@svenboulanger  Feel free to do it 👍
